File: src/algorithms/matrices/transformation/flip-image/sources/FlipImage_test.cpp
// g++ -std=c++17 -o flip_image_test FlipImage_test.cpp && ./flip_image_test
#include "FlipImage.cpp"
#include <cassert>
#include <iostream>

int main() {
    // test: flips and inverts 3x3 example
    {
        std::vector<std::vector<int>> matrix = {{1, 1, 0}, {1, 0, 1}, {0, 0, 0}};
        flipImage(matrix);
        assert((matrix[0] == std::vector<int>{1, 0, 0}));
        assert((matrix[1] == std::vector<int>{0, 1, 0}));
        assert((matrix[2] == std::vector<int>{1, 1, 1}));
    }

    // test: all zeros
    {
        std::vector<std::vector<int>> matrix = {{0, 0}, {0, 0}};
        flipImage(matrix);
        assert((matrix[0] == std::vector<int>{1, 1}));
        assert((matrix[1] == std::vector<int>{1, 1}));
    }

    // test: all ones
    {
        std::vector<std::vector<int>> matrix = {{1, 1}, {1, 1}};
        flipImage(matrix);
        assert((matrix[0] == std::vector<int>{0, 0}));
        assert((matrix[1] == std::vector<int>{0, 0}));
    }

    // test: single row
    {
        std::vector<std::vector<int>> matrix = {{1, 0, 1}};
        flipImage(matrix);
        assert((matrix[0] == std::vector<int>{0, 1, 0}));
    }

    // test: single column
    {
        std::vector<std::vector<int>> matrix = {{1}, {0}, {1}};
        flipImage(matrix);
        assert(matrix[0][0] == 0 && matrix[1][0] == 1 && matrix[2][0] == 0);
    }

    // test: 1x1 with 0
    {
        std::vector<std::vector<int>> matrix = {{0}};
        flipImage(matrix);
        assert(matrix[0][0] == 1);
    }

    // test: 1x1 with 1
    {
        std::vector<std::vector<int>> matrix = {{1}};
        flipImage(matrix);
        assert(matrix[0][0] == 0);
    }

    // test: identity-like matrix
    {
        std::vector<std::vector<int>> matrix = {{1, 0, 0}, {0, 1, 0}, {0, 0, 1}};
        flipImage(matrix);
        assert((matrix[0] == std::vector<int>{1, 1, 0}));
        assert((matrix[1] == std::vector<int>{1, 0, 1}));
        assert((matrix[2] == std::vector<int>{0, 1, 1}));
    }

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: src/algorithms/matrices/transformation/flip-image/__tests__/flip-image.test.ts
import { describe, it, expect } from "vitest";
import { flipImage } from "../sources/flip-image.ts?fn";

function deepCopy(matrix: number[][]): number[][] {
  return matrix.map((row) => [...row]);
}

describe("flipImage", () => {
  it("flips and inverts the default 3×3 example", () => {
    const matrix = deepCopy([
      [1, 1, 0],
      [1, 0, 1],
      [0, 0, 0],
    ]);
    expect(flipImage(matrix)).toEqual([
      [1, 0, 0],
      [0, 1, 0],
      [1, 1, 1],
    ]);
  });

  it("handles all-zeros matrix", () => {
    const matrix = deepCopy([
      [0, 0],
      [0, 0],
    ]);
    expect(flipImage(matrix)).toEqual([
      [1, 1],
      [1, 1],
    ]);
  });

  it("handles all-ones matrix", () => {
    const matrix = deepCopy([
      [1, 1],
      [1, 1],
    ]);
    expect(flipImage(matrix)).toEqual([
      [0, 0],
      [0, 0],
    ]);
  });

  it("handles a single row", () => {
    const matrix = deepCopy([[1, 0, 1]]);
    expect(flipImage(matrix)).toEqual([[0, 1, 0]]);
  });

  it("handles a single column", () => {
    const matrix = deepCopy([[1], [0], [1]]);
    expect(flipImage(matrix)).toEqual([[0], [1], [0]]);
  });

  it("handles a 1×1 matrix with value 0", () => {
    const matrix = deepCopy([[0]]);
    expect(flipImage(matrix)).toEqual([[1]]);
  });

  it("handles a 1×1 matrix with value 1", () => {
    const matrix = deepCopy([[1]]);
    expect(flipImage(matrix)).toEqual([[0]]);
  });

  it("handles a 4×4 binary matrix", () => {
    const matrix = deepCopy([
      [1, 1, 0, 0],
      [1, 0, 0, 1],
      [0, 1, 1, 0],
      [1, 0, 1, 0],
    ]);
    expect(flipImage(matrix)).toEqual([
      [1, 1, 0, 0],
      [0, 1, 1, 0],
      [1, 0, 0, 1],
      [1, 0, 1, 0],
    ]);
  });

  it("handles an identity-like binary matrix", () => {
    // Row 0: [1,0,0] → ends swap+invert: left=1→(0^1)=1, right=0→(1^1)=0, middle: 0→1 → [1,1,0]
    // Row 1: [0,1,0] → ends swap+invert: left=0→(0^1)=1, right=0→(0^1)=1, middle: 1→0 → [1,0,1]
    // Row 2: [0,0,1] → ends swap+invert: left=0→(1^1)=0, right=1→(0^1)=1, middle: 0→1 → [0,1,1]
    const matrix = deepCopy([
      [1, 0, 0],
      [0, 1, 0],
      [0, 0, 1],
    ]);
    expect(flipImage(matrix)).toEqual([
      [1, 1, 0],
      [1, 0, 1],
      [0, 1, 1],
    ]);
  });
});
